### PR TITLE
ref_use_double and eliminate_convert pass

### PR DESCRIPTION
**Pull request by CharlieL7**
* `ref` is used as our target values for `verify` tests. It should have higher accuracy than our GPU tests. Accuracy is muddled for fp16 values when comparing `gpu` to `ref`
* Adds new verify_option struct
* Can use driver flag `--ref-use-double` to convert all floating point on ref target to double precision.
* Creates a new eliminate_convert pass that handles more cases of nested converts.

### CODE REVIEW

File: src/driver/verify.cpp [hunk 1]
```diff
@@ -86,6 +88,7 @@ std::vector<argument> run_target(program p,
                                  precision quantize,
                                  const parameter_map& inputs)
 {
+    // NOTE: why do we have this here when driver/main.cpp:compiler already quantized
```
Comment (pfultz2) [inline src/driver/verify.cpp:91]: It doesn't quantize it for `verify`:

https://github.com/ROCm/AMDMIGraphX/blob/develop/src/driver/main.cpp#L579

So the ref is not quantized whereas the target is quantized.

File: src/driver/verify.cpp [hunk 2]
```diff
@@ -74,6 +75,7 @@ verify::tolerance get_tolerances(const program& p,
 
 std::vector<argument> run_ref(program p, const parameter_map& inputs)
 {
+    fp_to_double(p);
```
Comment (pfultz2) [inline src/driver/verify.cpp:78]: Do we want to do this unconditionally?

Comment (CharlieL7) [reply src/driver/verify.cpp:78]: We could add an option to the driver, but then need another flag to be passed.

**Comment by CharlieL7** [comment]
Figured out the loop_test bug. Was a logic mistake on this PR. This PR actually removes more nested converts now.
On develop loop_test we have this fused pointwise instruction:
```
loop_module:@10 = gpu::code_object[code_object=3168,symbol_name=equal_convert_convert_not_kernel,global=1,local=1024,](loop_module:@5,loop_module:loop_module:#output_0) -> bool_type, {1}, {0}, target_id=0
```
With this PR we get:
```
loop_module:@7 = gpu::code_object[code_object=3152,symbol_name=equal_convert_not_kernel,global=1,local=1024,](loop_module:@5,loop_module:loop_module:#output_0) -> bool_type, {1}, {0}, target_id=0
```

**Review by TedThemistokleous — APPROVED** [review]

File: src/include/migraphx/compile_options.hpp [hunk 3]
```diff
@@ -40,6 +40,12 @@ struct compile_options
 
     bool fast_math       = true;
     bool exhaustive_tune = false;
+
+    /**
+     * Converts floating point values to double on the ref target.
+     * Currently only used by the driver.
+     */
+    bool fp_to_double = false;
```
Comment (pfultz2) [inline src/include/migraphx/compile_options.hpp:48]: `compile_options` are for the `compile` function. If its not being used by `compile` then it shouldn't be there.

File: src/driver/main.cpp [hunk 4]
```diff
@@ -467,6 +467,10 @@ struct compiler
            {"--exhaustive-tune"},
            ap.help("Exhastively search for best tuning parameters for kernels"),
            ap.set_value(true));
+        ap(co.fp_to_double,
```
Comment (pfultz2) [inline src/driver/main.cpp:470]: This should go under `verify` not `compile` as its only used with `verify` command,

File: src/driver/main.cpp [hunk 5]
```diff
@@ -467,6 +467,10 @@ struct compiler
...
+           {"--fp-to-double"},
```
Comment (pfultz2) [inline src/driver/main.cpp:471]: This should also be named `--ref-use-double`.

File: src/include/migraphx/pass.hpp [hunk 6]
```diff
@@ -333,6 +333,21 @@ inline const ValueType& any_cast(const pass& x)
 
 #endif
 
+/// Dummy pass for default return
+struct id_pass
+{
+    std::string name() const { return "id"; }
+    void apple(const module&) const {}
```
Comment (pfultz2) [inline src/include/migraphx/pass.hpp:340]: It should be named `apply` not `apple`.

File: src/include/migraphx/pass.hpp [hunk 7]
```diff
@@ -333,6 +333,21 @@ inline const ValueType& any_cast(const pass& x)
...
+};
+
+/// Used in the targets to enable/disable compiler passes
+inline pass enable_pass(bool enabled, pass p)
```
Comment (pfultz2) [inline src/include/migraphx/pass.hpp:344]: This should be defined in a pass.cpp fle. And the `id_pass` class should be moved to the .cpp file as well.

File: src/include/migraphx/compile_options.hpp (hunk 3 above)
Comment (CharlieL7) [reply src/include/migraphx/compile_options.hpp:48]: The option is used to enable compiler passes, so it makes sense as a compile_option like disable_fast_math.

Comment (pfultz2) [reply src/include/migraphx/compile_options.hpp:48]: But we ignore the option for the GPU/CPU targets.

**Comment by pfultz2** [comment]
I think it would be better to create a `verify_options` struct that can store things like `precision` and whether ref will use `double`:

```cpp
struct verify_options
{
    precision quantize = precision::fp32;
    bool ref_use_double = false;
};
```

And you can pass that in place of the `quantize` parameter.

File: src/driver/verify.cpp [hunk 8]
```diff
@@ -72,9 +73,16 @@ verify::tolerance get_tolerances(const program& p,
...
+std::vector<argument> run_ref(program p,
+                              const compile_options& options,
+                              const verify_options& vo,
+                              const parameter_map& inputs)
 {
-    p.compile(migraphx::make_target("ref"));
+    p.compile(migraphx::make_target("ref"), options);
+    if(vo.ref_use_double)
+    {
+        run_passes(p, {fp_to_double{}});
```
Comment (pfultz2) [inline src/driver/verify.cpp:84]: Can we run this pass first before `compile`?

**Review by pfultz2 — APPROVED** [review]

**Comment by causten** [comment]
I can merged this with the performance numbers looking the way they are

**Comment by CharlieL7** [comment]
> I can merged this with the performance numbers looking the way they are

I don't see why this PR's changes would affect performance at all.

**Pull request merged**